src/custom/state/orders/mocks.ts

```diff
@@ -0,0 +1,180 @@
+    sellToken: sellToken.address.replace('0x', ''), // address, without '0x' prefix
+    buyToken: buyToken.address.replace('0x', ''), // address, without '0x' prefix
+    sellAmount, // in atoms
+    buyAmount, // in atoms
+    // 4 - 6min
+    validTo: Date.now() / 1000 + randomIntInRangeExcept(240, 360), // uint32. unix timestamp, seconds, use new Date(validTo * 1000)
+    appData: 1, // arbitrary identifier sent along with the order
+    feeAmount: (1e18).toString(), // in atoms
+    orderType,
+    partiallyFillable: false,
+    signature: (orderN++).toString().repeat(65 * 2) // 65 bytes encoded as hex without `0x` prefix. v + r + s from the spec
+  }
+}
+
+// add more orders if less than minPendingOrders currently
+const useAddOrdersOnMount = (minPendingOrders = 5) => {
+  const { account, chainId, library } = useActiveWeb3React()
+
+  const pendingOrders = usePendingOrders({ chainId })
+
+  // ref, so we don't rerun useEffect
+  const pendingOrdersRef = useRef(pendingOrders)
+  pendingOrdersRef.current = pendingOrders
+
+  const addOrder = useAddPendingOrder()
+
+  const lists = useSelectedTokenList()
+
+  useEffect(() => {
+    const addNOrders = (ordersNum: number) => {
+      if (!account || !chainId || !library) return
+      const tokenMap = lists[chainId]
+
+      const tokenList = Object.values(tokenMap)
+      if (tokenList.length === 0) return
+
+      const newTempOrders = Array.from({ length: ordersNum }, () => {
+        const [sellToken, buyToken] = getTwoRandomElementsFromArray(tokenList)
+
+        return generateOrder({
+          owner: account,
+          sellToken: sellToken.tokenInfo,
+          buyToken: buyToken.tokenInfo
+        })
+      })
+
+      batch(() => {
+        newTempOrders.forEach(order => {
+          addOrder({ order, id: order.id, chainId })
+        })
+      })
+    }
+    ;(window as any).addNOrders = addNOrders
+
+    // don't just keep adding orders when there's already enough pending
+    if (pendingOrdersRef.current.length >= minPendingOrders) return
+
+    addNOrders(10)
+  }, [account, addOrder, chainId, library, lists, minPendingOrders])
+}
+
+const useFulfillOrdersRandomly = (interval = 20000 /* ms */) => {
+  const { chainId } = useActiveWeb3React()
+  const pendingOrders = usePendingOrders({ chainId })
+
+  // ref, so we don't rerun useEffect
+  const pendingOrdersRef = useRef(pendingOrders)
+  pendingOrdersRef.current = pendingOrders
+
+  const fulfillOrder = useFulfillOrder()
+  const addPopup = useAddPopup()
+
+  useEffect(() => {
+    if (!chainId) return
+
+    const fulfillRandomOrder = () => {
+      // no more pending orders
+      // but don't clearInterval so we can restart when there are new orders
+      if (pendingOrdersRef.current.length === 0) return
+
+      const randomOrder = getRandomElementFromArray(pendingOrdersRef.current)
+
+      batch(() => {
+        fulfillOrder({ chainId, id: randomOrder.id, fulfillmentTime: new Date().toISOString() })
+
+        addPopup(
```

Comment:
@Velenir should this logic be done here?

The popup should be done the REAL one, right. When the order changes state, then there's a new popup showing. But it shouldn't be part of the mock

Reply:
Where would you put it?
Uniswap does this same thing in `transactions/updater`, updater sees an update -- shows a popup
I agree that a more centralized place would do better, but that would likely be in a middleware, like @W3stside did for fees.
Is that what you have in mind?

Reply:
Or you mean a separate `useEffect` that would monitor  when a previously `pending` order moves to `fulfilled`?
That would be ugly, would have to diff orders mappings
A much cleaner solution is a middleware

Actually I would happily make a middleware. After @W3stside merges the one he made for fees, so I don't get too many conflicts